services/graph/pkg/service/v0/links.go

```diff
@@ -0,0 +1,146 @@
+		return
+	}
+	if driveID.StorageId != driveItemID.StorageId || driveID.SpaceId != driveItemID.SpaceId {
+		errorcode.ItemNotFound.Render(w, r, http.StatusNotFound, "Item does not exist")
+		return
+	}
+	var createLink libregraph.DriveItemCreateLink
+	if err := StrictJSONUnmarshal(r.Body, &createLink); err != nil {
+		logger.Error().Err(err).Interface("body", r.Body).Msg("could not create link: invalid body schema definition")
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "invalid body schema definition")
+		return
+	}
+
+	createdLink, err := g.createLink(r.Context(), &driveItemID, createLink)
+	if err != nil {
+		errorcode.RenderError(w, r, err)
+		return
+	}
+
+	perm, err := g.libreGraphPermissionFromCS3PublicShare(createdLink)
+	if err != nil {
+		errorcode.GeneralException.Render(w, r, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	render.Status(r, http.StatusCreated)
+	render.JSON(w, r, []libregraph.Permission{*perm})
+}
+
+func (g Graph) createLink(ctx context.Context, driveItemID *providerv1beta1.ResourceId, createLink libregraph.DriveItemCreateLink) (*link.PublicShare, error) {
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Error().Err(err).Msg("could not select next gateway client")
+		return nil, errorcode.New(errorcode.GeneralException, err.Error())
+	}
+
+	statResp, err := gatewayClient.Stat(
+		ctx,
+		&providerv1beta1.StatRequest{
+			Ref: &providerv1beta1.Reference{
+				ResourceId: driveItemID,
+				Path:       ".",
+			},
+		})
+	if err != nil {
+		g.logger.Error().Err(err).Msg("transport error, could not stat resource")
+		return nil, errorcode.New(errorcode.GeneralException, err.Error())
+	}
+	if code := statResp.GetStatus().GetCode(); code != rpc.Code_CODE_OK {
+		g.logger.Debug().Interface("itemID", driveItemID).Msg(statResp.GetStatus().GetMessage())
+		return nil, errorcode.New(cs3StatusToErrCode(code), statResp.GetStatus().GetMessage())
+	}
+	permissions, err := linktype.CS3ResourcePermissionsFromSharingLink(createLink, statResp.GetInfo().GetType())
+	if err != nil {
+		g.logger.Debug().Interface("createLink", createLink).Msg(err.Error())
+		return nil, errorcode.New(errorcode.InvalidRequest, "invalid link type")
+	}
+	req := link.CreatePublicShareRequest{
+		ResourceInfo: statResp.GetInfo(),
+		Grant: &link.Grant{
+			Permissions: &link.PublicSharePermissions{
+				Permissions: permissions,
+			},
+			Password: createLink.GetPassword(),
+		},
+	}
+	if createLink.ExpirationDateTime != nil {
+		req.GetGrant().Expiration = utils.TimeToTS(createLink.GetExpirationDateTime())
+	}
+
+	// set displayname and password protected as arbitrary metadata
+	req.ResourceInfo.ArbitraryMetadata = &providerv1beta1.ArbitraryMetadata{
+		Metadata: map[string]string{
+			"name":      createLink.GetDisplayName(),
+			"quicklink": strconv.FormatBool(createLink.GetLibreGraphQuickLink()),
+		},
+	}
+	createResp, err := gatewayClient.CreatePublicShare(ctx, &req)
```

Comment:
errors with the following request body:

```json
{
  "type": "view"
}
```

`runtime error: invalid memory address or nil pointer dereference`

Reply:
turned out to be a problem with @fschade setup.

Reply:
@micbar fixed via https://github.com/cs3org/reva/pull/4365. the error still appears but now only because of the missing password (which is forced by default)... absolutely fine to have it there but needs a correct json error in those cases. 

if you don't like that reva refuses the request, we could add a validation for that too....
https://github.com/owncloud/ocis/blob/master/services/graph/pkg/validate/libregraph.go